File: examples/java/GateawayClient.java
package examples.gateaway;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.UUID;

public final class GateawayClient {
  private final HttpClient http = HttpClient.newHttpClient();
  private final String baseUrl;
  private final String apiKey;

  public GateawayClient(String baseUrl, String apiKey) {
    this.baseUrl = baseUrl;
    this.apiKey = apiKey;
  }

  public String createPaymentIntent(String orderReference, int amount) throws Exception {
    String json = "{\"amount\":" + amount + ",\"currency\":\"MZN\",\"orderReference\":\"" + orderReference + "\"}";
    return send("/v1/payment_intents", "POST", json, UUID.randomUUID().toString());
  }

  public String confirmWithMock(String paymentIntentId, String customerPhone) throws Exception {
    String json = "{\"paymentMethod\":\"mpesa\",\"customerPhone\":\"" + customerPhone + "\",\"provider\":\"mock\"}";
    return send("/v1/payment_intents/" + paymentIntentId + "/confirm", "POST", json, UUID.randomUUID().toString());
  }

  public String getPaymentIntent(String paymentIntentId) throws Exception {
    return send("/v1/payment_intents/" + paymentIntentId, "GET", null, null);
  }

  private String send(String path, String method, String body, String idempotencyKey) throws Exception {
    HttpRequest.Builder builder = HttpRequest.newBuilder(URI.create(baseUrl + path))
        .header("x-api-key", apiKey)
        .header("accept", "application/json");
    if (idempotencyKey != null) builder.header("idempotency-key", idempotencyKey);
    if (body != null) builder.header("content-type", "application/json");
    builder.method(method, body == null ? HttpRequest.BodyPublishers.noBody() : HttpRequest.BodyPublishers.ofString(body));
    HttpResponse<String> response = http.send(builder.build(), HttpResponse.BodyHandlers.ofString());
    if (response.statusCode() < 200 || response.statusCode() >= 300) {
      throw new IllegalStateException("GATEAWAY HTTP " + response.statusCode() + ": " + response.body());
    }
    return response.body();
  }
}

File: examples/typescript/gateaway-client.ts
import crypto from "crypto";

const baseUrl = process.env.GATEAWAY_BASE_URL ?? "http://localhost:3000";
const apiKey = required("GATEAWAY_API_KEY");

function required(name: string): string {
  const value = process.env[name];
  if (!value) throw new Error(`${name}_required`);
  return value;
}

async function request<T>(path: string, init: RequestInit): Promise<T> {
  const response = await fetch(`${baseUrl}${path}`, {
    ...init,
    headers: {
      "content-type": "application/json",
      "x-api-key": apiKey,
      ...(init.headers ?? {}),
    },
  });
  const body = await response.json();
  if (!response.ok) throw new Error(`${response.status}:${body.error ?? "request_failed"}`);
  return body as T;
}

export async function createPaymentIntent(orderReference: string, amount: number) {
  return request<{ paymentIntent: { id: string; status: string } }>("/v1/payment_intents", {
    method: "POST",
    headers: { "idempotency-key": crypto.randomUUID() },
    body: JSON.stringify({ amount, currency: "MZN", orderReference }),
  });
}

export async function confirmPaymentIntent(id: string, customerPhone: string) {
  return request<{ paymentIntent: { id: string; status: string } }>(`/v1/payment_intents/${id}/confirm`, {
    method: "POST",
    headers: { "idempotency-key": crypto.randomUUID() },
    body: JSON.stringify({ paymentMethod: "mpesa", customerPhone, provider: "mock" }),
  });
}

export async function getPaymentIntent(id: string) {
  return request<{ paymentIntent: { id: string; status: string } }>(`/v1/payment_intents/${id}`, { method: "GET" });
}
